### PR TITLE
Added zone redundant parameter to CLI commands

### PR DESCRIPTION
**Pull request by emgu-ms**
Added zone redundant parameter to the following CLI commands:
•sql db create
•sql db update
•sql elastic-pool create
•sql elastic-pool update

This allows customers to specify whether they want their database/elastic pool in a zone redundant ring or not. 

Wrote new tests targeting zone redundancy parameter and updated session records for them.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [X] Each command and parameter has a meaningful description.
- [X] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5560
This is an **experimental preview** for @microsoft users.

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 1]
```diff
@@ -16,14 +16,14 @@
...
-                 parameter_name_for_location='resource_group_location', location='westus',
+                 parameter_name_for_location='resource_group_location', location='eastus2',
                  dev_setting_name='AZURE_CLI_TEST_DEV_RESOURCE_GROUP_NAME',
                  dev_setting_location='AZURE_CLI_TEST_DEV_RESOURCE_GROUP_LOCATION',
                  random_name_length=75, key='rg'):
         super(ResourceGroupPreparer, self).__init__(name_prefix, random_name_length)
         from azure.cli.testsdk import TestCli
         self.cli_ctx = TestCli()
-        self.location = location
+        self.location = "eastus2"
```
Comment (jaredmoo) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: this change should not be needed

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py [hunk 2]
```diff
@@ -115,9 +115,6 @@ def _configure_db_create_params(
     create_mode: Valid CreateMode enum value (e.g. `default`, `copy`, etc)
     """
 
-    # This line to be removed when zone redundancy is fully supported in production.
-    arg_ctx.ignore('zone_redundant')
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py]: is zone_redundant allowed for all create modes (copy, geo secondary, restore)? is it allowed for DW?

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py [hunk 3]
```diff
@@ -53,10 +53,20 @@ def _get_server_location(cli_ctx, server_name, resource_group_name):
...
 
 
+# Returns True/False boolean given an input boolean in string 'true'/'false' format regardless of casing
+def str_to_bool(input_str):
+    if input_str.lower() == 'true':
+        return True
+    elif input_str.lower() == 'false':
+        return False
+    else:
+        raise ValueError("Cannot covert {} to a bool".format(input_str))
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py]: Why is this needed? I thought 3 state flag should convert to bool automatically

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 4]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+        self.cmd('sql db create -g {} --server {} --name {} --edition {}'
+                 .format(rg, server, database_name_3, "Premium"),
+                 checks=[
+                     JMESPathCheck('resourceGroup', rg),
+                     JMESPathCheck('name', database_name_3),
+                     JMESPathCheck('location', loc_display),
+                     JMESPathCheck('elasticPoolName', None),
+                     JMESPathCheck('status', 'Online'),
+                     JMESPathCheck('edition', 'Premium'),
+                     JMESPathCheck('zoneRedundant', False)]).get_output_in_json()
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: You can just add `JMESPathCheck('zoneRedundant', False)` to the existing create db test.

Comment (jaredmoo) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: This database_name_3 case is not needed.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 5]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+        database_name = "updateToZonedDb"
+        database_name_2 = "updateToUnzonedDb"
+        database_name_3 = "updateNoParamForUnzonedDb"
+        database_name_4 = "updateNoParamForZonedDb"
+
+        rg = resource_group
+        loc_display = "East US 2"
+
+        # Test running update on regular database with zone resilience set to true.  Expect zone resilience to update to true.
+        self.cmd('sql db create -g {} --server {} --name {} --edition {}'
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: This is already the 1st step in the "create zone redundant" test case, so you can combine the 2 test cases to reduce test e2e run time.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 6]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+                 .format(rg, server, pool_name_3, "Premium"))
+
+        self.cmd('sql elastic-pool show -g {} --server {} --name {}'
+                 .format(rg, server, pool_name_3),
+                 checks=[
+                     JMESPathCheck('resourceGroup', rg),
+                     JMESPathCheck('name', pool_name_3),
+                     JMESPathCheck('state', 'Ready'),
+                     JMESPathCheck('edition', 'Premium'),
+                     JMESPathCheck('zoneRedundant', False)]).get_output_in_json()
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: Same comment as db - this check can be in basic pool test

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 7]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+    @SqlServerPreparer()
+    def test_sql_update_zone_resilient_pool(self, resource_group, resource_group_location, server):
+        pool_name = "updateToZonedPool"
+        pool_name_2 = "updateToUnzonedPool"
+        pool_name_3 = "updateNoParamForUnzonedPool"
+        pool_name_4 = "updateNoParamForZonedPool"
+
+        rg = resource_group
+
+        # Test running update on regular pool with zone resilience set to true.  Expect zone resilience to update to true
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: This comment does not match the next line of code

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 8]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+                 .format(rg, server, pool_name, "Premium"))
+
+        self.cmd('sql elastic-pool show -g {} --server {} --name {}'
+                 .format(rg, server, pool_name),
+                 checks=[
+                     JMESPathCheck('resourceGroup', rg),
+                     JMESPathCheck('name', pool_name),
+                     JMESPathCheck('state', 'Ready'),
+                     JMESPathCheck('edition', 'Premium'),
+                     JMESPathCheck('zoneRedundant', False)]).get_output_in_json()
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: `JMESPathCheck('zoneRedundant', False)` can be in basic pool test

Comment (jaredmoo) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: I see you using `get_output_in_json()` a lot, this is only needed if you actually need to access any of the resource's properties. In most places you don't need it

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 9]
```diff
@@ -1939,3 +1939,300 @@ def test_sql_subscription_usages(self):
...
+
+        self.cmd('sql elastic-pool update -g {} -s {} -n {} --dtu {} --zone-redundant {}'
+                 .format(rg, server, pool_name, 125, True))
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: DTU seems irrelevant, can you remove it here?

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py (hunk 2 above)
Comment (emgu-ms) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py]: Nope, so I had to block the different create modes and dw

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py [hunk 10]
```diff
@@ -204,6 +201,11 @@ def load_arguments(self, _):
                    options_list=['--edition'],
                    help='The edition of the database.')
 
+        c.argument('zone_redundant',
+                   options_list=['--zone-redundant', '-z'],
+                   help='The zone redundancy to associate with the database.',
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py]: This description sounds a bit weird to me. How about something like "Specifies whether to enable zone redundancy for the database." And then please add examples to help.py like this:

```
- Create database with zone redundancy enabled
- az sql db create -g ... -s ... -n ... -z

- Create database with zone redundancy explicitly disabled
- az sql db create -g ... -s ... -n ... -z false
```

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py [hunk 11]
```diff
@@ -560,6 +559,11 @@ def _configure_security_policy_storage_params(arg_ctx):
                    help='The max storage size of the elastic pool. If no unit is specified, defaults'
                    ' to megabytes (MB).')
 
+        c.argument('zone_redundant',
+                   options_list=['--zone-redundant', '-z'],
+                   help='The zone redundancy to associate with the elastic pool.',
```
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py]: Same comment about the help text.
